x11rb-protocol/src/parse_display/connect_instruction.rs

```diff
@@ -0,0 +1,218 @@
+//! Provides the `ConnectInstruction` structure, which allows for a `ParsedDisplay`
+//! to be transformed into a server connection.
+
+use super::ParsedDisplay;
+use std::{iter::FusedIterator, path::PathBuf};
+
+const TCP_PORT_BASE: u16 = 6000;
+
+/// Instructions for connecting to the X11 server, such as what type of socket to
+/// use and what address to connect to.
+///
+/// This structure allows for a `ParsedDisplay` to be transformed into a server
+/// connection.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ConnectInstruction<'a> {
+    parsed_display: &'a ParsedDisplay,
+    stage: Stage,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+enum Stage {
+    TryHostname,
+    TrySocket,
+    TryPort,
+    OutOfOptions,
+}
+
+impl Default for Stage {
+    fn default() -> Self {
+        Stage::TryHostname
+    }
+}
+
+/// Either a hostname and port to connect to, or a socket to connect to.
+// Do we want to make this #[non_exhaustive], in case we add more connection
+// instructions in the future? i.e. what if we want to do X11 over QUIC?
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ConnectAddress<'a> {
+    /// Connect to this hostname and port over TCP.
+    Hostname(&'a str, u16),
+    /// Connect to this Unix socket.
+    Socket(PathBuf),
+}
+
+impl<'a> ConnectInstruction<'a> {
+    #[inline]
+    pub(super) fn from_parsed_display(parsed_display: &'a ParsedDisplay) -> Self {
+        ConnectInstruction {
+            parsed_display,
+            stage: Stage::default(),
+        }
+    }
+
+    /// We should try to connect to the hostname and port over TCP.
+    fn try_hostname_over_tcp(&self) -> bool {
+        let ParsedDisplay { host, protocol, .. } = self.parsed_display;
+
+        !matches!(
+            (host.as_str(), protocol.as_deref()),
+            ("", _) | ("unix", _) | (_, Some("unix"))
+        )
+    }
+
+    /// We should try to connect to a socket.
+    fn try_unix_socket(&self) -> bool {
+        let ParsedDisplay { protocol, .. } = self.parsed_display;
+
+        cfg!(unix) && matches!(protocol.as_deref(), None | Some("unix"))
```

Comment:
I think this `cfg!(unix)` can be left out. x11rb-protocol does not do I/O and thus does not need to care whether it is on Windows or Unix. The API user can skip the unix connection attempts (and has to do so anyway since already in this version of the PR `ConnectAddress::Socket` always exists).

Reply:
In that case, I think `String` should be used instead of `PathBuf`, as the later type is platform-dependent.